pennylane/templates/subroutines/select.py

```diff
@@ -228,3 +233,230 @@ def target_wires(self):
     def wires(self):
         """All wires involved in the operation."""
         return self.hyperparameters["control"] + self.hyperparameters["target_wires"]
+
+
+def _ceil(a):
+    return int(math.ceil(a))
+
+
+def _ceil_log(a):
+    return _ceil(math.log2(a))
+
+
+def _add_first_k_units(ops, controls, work_wires, k):
+    """Add all controlled-applied operators within the unary iterator scheme.
+
+    This function is used for the outer-most recursion level, and then calls _add_k_units
+    for the inner recursion levels.
+
+    Given ``k=len(ops)=2**a-b`` operators, where ``a`` is chosen as small as possible to obtain
+    ``0<=b<2**(a-1)`` (note that this implies ``a=⌈log_2(k)⌉``), and ``2 * a - 1`` control wires,
+    this function applies one of three circuits; in each variant, the first ``2**(a-1)`` are
+    applied in two equal portions of ``2**(a-2)`` operators, after which ``l=2**(a-1) -b``
+    operators remain. Now the three circuit variants are distinguished, based on ``l``:
+
+    - if ``l=1``, the following circuit is applied:
+      ```
+      ─╭○─────╭○─────○─╮╭●──
+      ─├○─────│──────●─┤│───
+       ╰──╭■──╰X─╭■────╯│───
+      ────├■─────├■─────│───
+          ╰■     ╰■     ╰■  .
+      ```
+      Here, each triple box symbolizes a call to ``_add_k_units`` to add 2**(a-2) operators in a
+      recursive manner, and the right hand side controlled gate is a single controlled operator.
+
+    - if ``l < 2**(a-2)``, the following circuit is applied:
+      ```
+      ─╭○─────╭○─────○─╮╭●─────╭●─────●─╮─
+      ─├○─────│──────●─┤│──────│────────│─
+       ╰──╭■──╰X─╭■────╯│      │        │
+      ────├■─────├■─────├○─────│──────●─┤─
+          ╰■     ╰■     ╰───■──╰X──■────╯
+      ```
+      where the second half may skip more than two control wires (the number of skipped control
+      wires is ``c_bar`` in code). The single boxes in the second half are calls to
+      ``_add_k_units``, not single operators.
+
+    - if ``l >= 2**(a-2)``, the following, more generic, circuit is applied:
+      ```
+      ─╭○─────╭○─────╭●────────╭●─────●─╮─
+      ─├○─────│──────│──╭●─────│──────●─┤─
+       ╰──╭■──╰X─╭■──╰X─╰X─╭■──╰X─╭■────╯
+      ────├■─────├■────────├■─────├■──────
+          ╰■     ╰■        ╰■     ╰■      .
```

Comment:
How do you choose the number of operators that goes to each half of the second block? By saturating the left one first?

Reply:
Added a comment to this and the case above.